File: kiwa-go/integration.go
// Package kiwa: integration test helper.
//
// MockServer wraps net/http/httptest.Server with the kiwa fixture contract:
// route table evaluated in registration order, a request recorder that
// captures every request the server sees (matched or not), and automatic
// shutdown via testing.TB.Cleanup so tests cannot forget to release the
// underlying TCP port.
//
// It mirrors the kiwa-test-rs integration::mock_server API 1:1 — same Route
// builder, same MockResponse shape, same recorder semantics — so a multi-
// language Layer 1 spec compiles to either ecosystem without diverging.
//
// # Differentiation vs raw net/http/httptest
//
// httptest.NewServer is already a thin and correct wrapper; kiwa.MockServer
// does not try to hide it. The value added is purely contract:
//
//   - Route table (method + exact path → handler) so a test reads as data
//     rather than as a switch-case wired into a single http.HandlerFunc.
//   - Recorder snapshot — RecordedRequests returns the request log as
//     []RecordedRequest with method / path / headers / body fully captured,
//     including for unmatched routes (which httptest leaves to the caller).
//   - testing.TB.Cleanup integration so port release matches the unit
//     fixture lifecycle (SetupUnitEnv) — one mental model for both layers.
//
// For request matchers richer than "exact method + exact path" (regex, JSON
// path, response sequencing) reach for github.com/h2non/gock or
// github.com/jarcoal/httpmock — both coexist with kiwa fixtures.
package kiwa

import (
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"

	"github.com/cardene777/kiwa-test-go/internal/recorder"
)

// HTTPMethod is a Route's HTTP method, expressed as a kiwa-owned enum so test
// code does not have to import net/http just to pick GET vs POST.
type HTTPMethod int

const (
	// MethodGET matches GET requests.
	MethodGET HTTPMethod = iota
	// MethodPOST matches POST requests.
	MethodPOST
	// MethodPUT matches PUT requests.
	MethodPUT
	// MethodPATCH matches PATCH requests.
	MethodPATCH
	// MethodDELETE matches DELETE requests.
	MethodDELETE
	// MethodHEAD matches HEAD requests.
	MethodHEAD
	// MethodOPTIONS matches OPTIONS requests.
	MethodOPTIONS
)

// String returns the canonical HTTP method label (e.g. "GET").
func (m HTTPMethod) String() string {
	switch m {
	case MethodGET:
		return http.MethodGet
	case MethodPOST:
		return http.MethodPost
	case MethodPUT:
		return http.MethodPut
	case MethodPATCH:
		return http.MethodPatch
	case MethodDELETE:
		return http.MethodDelete
	case MethodHEAD:
		return http.MethodHead
	case MethodOPTIONS:
		return http.MethodOptions
	default:
		return "UNKNOWN"
	}
}

// matches reports whether this method enum equals the wire-level HTTP method
// string carried on an *http.Request.
func (m HTTPMethod) matches(wireMethod string) bool {
	return m.String() == wireMethod
}

// RecordedRequest is a snapshot of a single inbound HTTP request observed by
// the mock server.
//
// The struct definition lives in the internal recorder package
// (v1.6-5 dedup, Issue #611) — this alias keeps the pre-v1.6-5 name on
// the exported kiwa surface so existing callers do not have to change
// their imports. Both header views (single-value Headers, multi-value
// HeadersAll) and the always-defensively-copied Body semantics are
// documented on the underlying recorder.Snapshot type.
type RecordedRequest = recorder.Snapshot

// MockResponse is the response a Route handler returns.
//
// The zero value is a 200 OK with no headers and an empty body — matching
// httptest defaults so callers only spell out the fields they care about.
//
// Two header views coexist: Headers writes a single value per key (mirroring
// the pre-v1.6 shape), HeadersAll writes every value in wire order so
// multi-value headers (Set-Cookie, WWW-Authenticate, Vary, Link, ...) can be
// emitted verbatim. writeResponse merges the two — HeadersAll entries add to
// the response, then Headers entries Set (replacing anything HeadersAll wrote
// for that key) so callers who only fill Headers keep the previous behaviour.
type MockResponse struct {
	// Status is the HTTP status code. Zero is treated as 200.
	Status int
	// Headers are the response headers (mixed-case keys preserved; net/http
	// normalises them on the wire). Single value per key.
	Headers map[string]string
	// HeadersAll are the response headers when multiple values per key are
	// required (Set-Cookie in particular). Values are written in slice order
	// via http.Header.Add so the wire keeps every entry.
	HeadersAll map[string][]string
	// Body is the raw response body bytes.
	Body []byte
}

// WithHeaderValues returns a copy of r with every value in values appended
// to HeadersAll[key] (the existing HeadersAll map is cloned so callers can
// chain safely). Use this for Set-Cookie / WWW-Authenticate / Vary / Link
// where multiple header lines matter.
func (r MockResponse) WithHeaderValues(key string, values ...string) MockResponse {
	cloned := make(map[string][]string, len(r.HeadersAll)+1)
	for k, v := range r.HeadersAll {
		copied := make([]string, len(v))
		copy(copied, v)
		cloned[k] = copied
	}
	// Append instead of overwriting so the caller can chain WithHeaderValues
	// per line — matches http.Header.Add semantics.
	appended := make([]string, 0, len(cloned[key])+len(values))
	appended = append(appended, cloned[key]...)
	appended = append(appended, values...)
	cloned[key] = appended
	r.HeadersAll = cloned
	return r
}

// OK builds a 200 OK response with the given body.
//
// The body slice is defensively copied at ingress so route handlers
// scratch-buffer reuse (a common pattern when a handler formats the
// response body into a per-request bytes.Buffer or an sync.Pool-backed
// slice) cannot retroactively mutate the wire payload after OK
// returns. v1.6-2 (Issue #608) hazard 1.
func OK(body []byte) MockResponse {
	return MockResponse{Status: http.StatusOK, Body: cloneBody(body)}
}

// JSON builds a 200 OK response with Content-Type: application/json. The
// caller serialises the value; we merely set the header.
//
// Same ingress defensive copy discipline as OK — see the OK godoc.
func JSON(body []byte) MockResponse {
	return MockResponse{
		Status:  http.StatusOK,
		Headers: map[string]string{"Content-Type": "application/json"},
		Body:    cloneBody(body),
	}
}

// WithStatus returns a copy of r with the status overridden.
func (r MockResponse) WithStatus(status int) MockResponse {
	r.Status = status
	return r
}

// WithBody returns a copy of r with the body replaced. The body slice
// is defensively copied at ingress so caller scratch-buffer reuse
// cannot retroactively mutate the wire payload after WithBody returns
// — same discipline as OK / JSON (v1.6-2 hazard 1). Callers who prefer
// to spell out the response as a struct literal can still do so; those
// paths will get ingress copy discipline when the response is written
// (see writeResponse and the defensive copy in newMockResponse below).
func (r MockResponse) WithBody(body []byte) MockResponse {
	r.Body = cloneBody(body)
	return r
}

// cloneBody defensively copies a response body slice at MockResponse
// ingress so caller-side buffer reuse (typically a bytes.Buffer or
// sync.Pool-backed scratch slice inside the route handler) cannot
// retroactively mutate the wire payload. Returns a fresh []byte even
// for a nil input so downstream serialisation code can iterate without
// a nil check. v1.6-2 (Issue #608) hazard 1.
func cloneBody(body []byte) []byte {
	out := make([]byte, len(body))
	copy(out, body)
	return out
}

// WithHeader returns a copy of r with header (key, value) inserted. The
// existing Headers map is cloned so callers can safely chain.
func (r MockResponse) WithHeader(key, value string) MockResponse {
	cloned := make(map[string]string, len(r.Headers)+1)
	for k, v := range r.Headers {
		cloned[k] = v
	}
	cloned[key] = value
	r.Headers = cloned
	return r
}

// RouteHandler is the closure type a Route invokes when its (method, path)
// matches an incoming request. The handler receives the recorded snapshot
// (already appended to the recorder before the handler runs) so it can
// inspect headers / body without re-reading the request.
type RouteHandler func(req RecordedRequest) MockResponse

// Route binds (method, path) to a RouteHandler.
//
// Path matching is exact for v0.1 (no glob, no regex); the route table is
// scanned in registration order and the first match wins. This keeps the
// matcher dead-simple and easy to reason about — promotion to richer
// matchers is a v0.2 candidate behind an opt-in builder method.
type Route struct {
	Method  HTTPMethod
	Path    string
	Handler RouteHandler
}

// NewRoute builds a Route from (method, path, handler).
func NewRoute(method HTTPMethod, path string, handler RouteHandler) Route {
	return Route{Method: method, Path: path, Handler: handler}
}

// MockServerOpts configures MockServer.
//
// The zero value is a server with no routes registered — every request 404s
// but is still recorded, which is the right default for tests that only care
// about whether the production code under test fired any HTTP traffic at all.
type MockServerOpts struct {
	// Routes is the route table evaluated in registration order.
	Routes []Route
}

// WithRoute appends route and returns the updated opts (chainable builder
// for call-site readability).
func (o MockServerOpts) WithRoute(route Route) MockServerOpts {
	o.Routes = append(o.Routes, route)
	return o
}

// MockServer is the running mock server handle returned by NewMockServer.
//
// Both Stop and the recorder are safe for concurrent use — Stop guards the
// underlying httptest.Server.Close with a mutex so manual Stop + t.Cleanup
// + parallel test bodies never race, and the recorder protects the
// captured-request log so the server goroutines dispatching incoming
// requests cannot tear writes against RecordedRequests / RequestCount
// reads from the test goroutine.
type MockServer struct {
	server   *httptest.Server
	recorder *recorder.Log
	stopped  bool
	stopMu   sync.Mutex
}

// URL returns the base URL clients should target (e.g. "http://127.0.0.1:54321").
func (s *MockServer) URL() string {
	return s.server.URL
}

// RecordedRequests returns a snapshot of every request the server has seen
// so far. The slice is freshly allocated and every Body / HeadersAll value
// is deep-copied so callers may iterate and mutate without racing the
// server goroutines that own the underlying log — parity with kiwa-rs
// Vec<u8>::clone (v1.6-2 hazard 2).
func (s *MockServer) RecordedRequests() []RecordedRequest {
	return s.recorder.Snapshot()
}

// RequestCount reports how many requests the server has received.
func (s *MockServer) RequestCount() int {
	return s.recorder.Count()
}

// Stop releases the underlying httptest.Server. Idempotent — t.Cleanup
// invokes this too, so an explicit Stop followed by cleanup is safe.
func (s *MockServer) Stop() {
	s.stopMu.Lock()
	defer s.stopMu.Unlock()
	if s.stopped {
		return
	}
	s.stopped = true
	s.server.Close()
}

// NewMockServer spins up an httptest.Server bound to a fresh ephemeral port
// and registers t.Cleanup to release the port when t finishes.
//
// NewMockServer accepts testing.TB so it works inside *testing.T,
// *testing.B, and *testing.F bodies, and calls t.Helper so failure stack
// frames point at the caller.
//
// # Example
//
//	srv := kiwa.NewMockServer(t, kiwa.MockServerOpts{}.WithRoute(
//	    kiwa.NewRoute(kiwa.MethodGET, "/users", func(_ kiwa.RecordedRequest) kiwa.MockResponse {
//	        return kiwa.JSON([]byte(`[{"id":1,"name":"sora"}]`))
//	    }),
//	))
//
//	resp, _ := http.Get(srv.URL() + "/users")
//	// ... assert on resp / srv.RecordedRequests() ...
func NewMockServer(t testing.TB, opts MockServerOpts) *MockServer {
	t.Helper()

	log := &recorder.Log{}
	// Clone the route table so caller-side mutation of opts.Routes after
	// NewMockServer returns does not affect server behaviour.
	routes := make([]Route, len(opts.Routes))
	copy(routes, opts.Routes)

	handler := buildHandler(log, routes)
	server := httptest.NewServer(handler)

	mock := &MockServer{server: server, recorder: log}
	t.Cleanup(mock.Stop)
	return mock
}

// buildHandler wires the recorder + route table into a single http.Handler.
// Factored out so unit tests can exercise the dispatcher without binding a
// real port (currently only end-to-end tests exercise it; the factoring
// keeps the option open).
//
// Request-to-Snapshot conversion (header case-folding, body defensive
// copy, path+query composition) is delegated to internal/recorder.FromServer
// — the single SSOT shared with kiwa/gin and kiwa/echo (v1.6-5, Issue #611).
func buildHandler(log *recorder.Log, routes []Route) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		// Capture the request before invoking the handler so the recorder
		// reflects observed traffic even if the handler panics or 404s.
		recorded := recorder.FromServer(req)
		log.Append(recorded)

		for _, route := range routes {
			if route.Method.matches(req.Method) && route.Path == req.URL.Path {
				writeResponse(w, route.Handler(recorded))
				return
			}
		}

		// No route matched — surface a self-describing 404 so accidental
		// rogue requests fail the test rather than hang the caller.
		body := fmt.Sprintf("kiwa mock_server: no route matched %s %s", req.Method, req.URL.Path)
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.WriteHeader(http.StatusNotFound)
		_, _ = io.WriteString(w, body)
	})
}

// writeResponse serialises a MockResponse onto the http.ResponseWriter.
// Headers must be written before WriteHeader; body comes last.
//
// HeadersAll is added first (Add per value so multi-line headers like
// Set-Cookie survive on the wire). Headers is applied afterwards with Set,
// which replaces any HeadersAll entry for the same key — so callers who only
// fill Headers keep the pre-v1.6 last-write-wins behaviour, and callers who
// only fill HeadersAll get every value on the wire in slice order.
func writeResponse(w http.ResponseWriter, resp MockResponse) {
	for k, values := range resp.HeadersAll {
		for _, v := range values {
			w.Header().Add(k, v)
		}
	}
	for k, v := range resp.Headers {
		w.Header().Set(k, v)
	}
	status := resp.Status
	if status == 0 {
		status = http.StatusOK
	}
	w.WriteHeader(status)
	if len(resp.Body) > 0 {
		_, _ = w.Write(resp.Body)
	}
}

File: kiwa-rs/src/integration.rs
//! Integration test helpers — hyper-based mock server + request recorder.
//!
//! This module wraps [`hyper`] behind a synchronous, kiwa-flavoured API so a
//! cargo test can spin up an HTTP endpoint, point a `reqwest` / `hyper` client
//! at it, and assert on the recorded requests — without leaking ports, without
//! a tokio runtime in the test, and without forgetting cleanup.
//!
//! ## Contract
//!
//! - [`mock_server`] returns a [`MockServer`] that owns a background tokio
//!   runtime and a single-thread server task. The runtime + task are torn down
//!   in [`Drop`], so each test releases its port deterministically.
//! - [`MockServerOpts`] is the input: optional fixed port (defaults to an
//!   OS-assigned ephemeral port), optional response timeout, and a list of
//!   [`Route`] handlers matched in registration order.
//! - [`Route`] couples (method, path) to a closure `Fn(&RecordedRequest) -> MockResponse`.
//!   Path matching is exact for v0.1 (no glob / regex); the response body is
//!   `Vec<u8>` so JSON / binary fixtures share one code path.
//! - Every incoming request is appended to a recorder regardless of whether a
//!   route matched. [`MockServer::recorded_requests`] returns a snapshot the
//!   test can iterate without touching internal locks.
//! - If no route matches the recorder still captures the request and the
//!   server replies with `404 Not Found`, so unexpected client behaviour
//!   surfaces as a test failure rather than a hang.
//!
//! ## Differentiation vs `httpmock` / `wiremock-rs`
//!
//! `kiwa::integration::mock_server` is intentionally minimal: same fixture
//! contract as the rest of kiwa (`setup_env` + Drop cleanup), exact path match,
//! sync handlers. For richer matchers (regex / JSON-path / response sequence),
//! reach for [`httpmock`] or [`wiremock`] — they coexist with kiwa fixtures.
//! See the crate README for the full comparison table.

use std::collections::HashMap;
use std::convert::Infallible;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use bytes::Bytes;
use http::{Method, StatusCode};
use http_body_util::{BodyExt, Full};
use hyper::body::Incoming;
use hyper::server::conn::http1;
use hyper::service::service_fn;
use hyper::{Request, Response};
use hyper_util::rt::TokioIo;
use tokio::net::TcpListener;
use tokio::runtime::Runtime;
use tokio::sync::oneshot;
use tokio::task::JoinHandle;

/// HTTP method on a [`Route`].
///
/// Re-exported as a kiwa-owned enum so test code does not need to depend on
/// `http::Method` directly — but matches that crate's semantics 1:1.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
#[allow(missing_docs)]
pub enum HttpMethod {
    Get,
    Post,
    Put,
    Patch,
    Delete,
    Head,
    Options,
}

impl HttpMethod {
    fn matches(self, m: &Method) -> bool {
        match self {
            HttpMethod::Get => m == Method::GET,
            HttpMethod::Post => m == Method::POST,
            HttpMethod::Put => m == Method::PUT,
            HttpMethod::Patch => m == Method::PATCH,
            HttpMethod::Delete => m == Method::DELETE,
            HttpMethod::Head => m == Method::HEAD,
            HttpMethod::Options => m == Method::OPTIONS,
        }
    }
}

/// Snapshot of an HTTP request received by the mock server.
///
/// The recorder stores a `Vec<RecordedRequest>` and hands the test a clone so
/// assertions cannot race with the server task.
///
/// Two header views coexist so multi-value headers (`Set-Cookie`,
/// `WWW-Authenticate`, `Vary`, `Link`, …) survive the recording round trip
/// while ergonomic single-value access stays cheap:
///
/// - [`headers`](Self::headers) — last-write-wins `HashMap<String, String>`,
///   retained for backward compatibility with pre-v1.6 assertions.
/// - [`headers_all`](Self::headers_all) — `HashMap<String, Vec<String>>` that
///   preserves every recorded value in wire order.
#[derive(Clone, Debug)]
pub struct RecordedRequest {
    /// Request method (`GET` / `POST` / …).
    pub method: String,
    /// Request path including query (e.g. `/users?id=1`).
    pub path: String,
    /// Request headers (lowercased keys, last-write-wins on duplicates).
    /// Retained for backward compatibility — new assertions that need
    /// multi-value semantics should read [`headers_all`](Self::headers_all).
    pub headers: HashMap<String, String>,
    /// Request headers (lowercased keys) preserving every value in wire order.
    /// Populated alongside [`headers`](Self::headers) so callers can pick the
    /// shape that fits without paying a second recorder pass.
    pub headers_all: HashMap<String, Vec<String>>,
    /// Request body bytes (empty `Vec` if the client sent no body).
    pub body: Vec<u8>,
}

impl RecordedRequest {
    /// Convenience — returns the body interpreted as UTF-8 (lossy).
    pub fn body_str(&self) -> String {
        String::from_utf8_lossy(&self.body).into_owned()
    }

    /// Return every recorded value for `key` (case-insensitive) in wire
    /// order. Returns `None` when the header was not observed so callers can
    /// distinguish "absent" from "present but empty".
    pub fn headers_all_values(&self, key: &str) -> Option<Vec<String>> {
        self.headers_all.get(&key.to_lowercase()).cloned()
    }
}

/// Response a [`Route`] handler returns.
///
/// Two header views coexist: [`headers`](Self::headers) writes a single value
/// per key (mirroring the pre-v1.6 shape), [`headers_all`](Self::headers_all)
/// writes every value in wire order so multi-value headers (`Set-Cookie`,
/// `WWW-Authenticate`, `Vary`, `Link`, …) can be emitted verbatim.
/// [`build_response`] merges the two — `headers_all` entries are added first
/// (append per line), then `headers` entries via `header()` add — matching
/// `http::Response::builder` semantics and preserving the pre-v1.6 behaviour
/// for callers that only fill `headers`.
#[derive(Clone, Debug)]
pub struct MockResponse {
    /// HTTP status code (defaults to `200`).
    pub status: u16,
    /// Response headers (single value per key).
    pub headers: HashMap<String, String>,
    /// Response headers when multiple values per key are required
    /// (`Set-Cookie` in particular). Values are appended in slice order via
    /// `http::Response::builder().header()` so the wire keeps every entry.
    pub headers_all: HashMap<String, Vec<String>>,
    /// Response body bytes.
    pub body: Vec<u8>,
}

impl Default for MockResponse {
    fn default() -> Self {
        MockResponse {
            status: 200,
            headers: HashMap::new(),
            headers_all: HashMap::new(),
            body: Vec::new(),
        }
    }
}

impl MockResponse {
    /// Build a `200 OK` response with the given body.
    pub fn ok(body: impl Into<Vec<u8>>) -> Self {
        MockResponse {
            status: 200,
            headers: HashMap::new(),
            headers_all: HashMap::new(),
            body: body.into(),
        }
    }

    /// Build a `200 OK` JSON response. The caller serialises the value; we
    /// merely set `Content-Type: application/json`.
    pub fn json(body: impl Into<Vec<u8>>) -> Self {
        let mut headers = HashMap::new();
        headers.insert("content-type".into(), "application/json".into());
        MockResponse {
            status: 200,
            headers,
            headers_all: HashMap::new(),
            body: body.into(),
        }
    }

    /// Override the status code.
    pub fn with_status(mut self, status: u16) -> Self {
        self.status = status;
        self
    }

    /// Insert a response header (single value; overrides any prior value for
    /// this key on the wire).
    pub fn with_header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(key.into().to_lowercase(), value.into());
        self
    }

    /// Append every value in `values` to the multi-value slot for `key`
    /// (case-insensitive). Chain multiple calls to emit multiple `Set-Cookie`
    /// / `WWW-Authenticate` / `Vary` / `Link` lines on the wire. Matches
    /// `http::HeaderMap::append` semantics.
    pub fn with_header_values<I, S>(mut self, key: impl Into<String>, values: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        let key = key.into().to_lowercase();
        let entry = self.headers_all.entry(key).or_default();
        for v in values {
            entry.push(v.into());
        }
        self
    }
}

/// Route handler — `(method, path)` mapped to a closure that produces a
/// [`MockResponse`].
///
/// Handlers are `Fn` (not `FnMut`) so they can run on the server task without
/// `&mut` aliasing; capture `Arc<Mutex<…>>` for stateful behaviour.
pub struct Route {
    method: HttpMethod,
    path: String,
    handler: Arc<dyn Fn(&RecordedRequest) -> MockResponse + Send + Sync + 'static>,
}

impl Route {
    /// Build a route from `(method, path)` and a handler closure.
    pub fn new<F>(method: HttpMethod, path: impl Into<String>, handler: F) -> Self
    where
        F: Fn(&RecordedRequest) -> MockResponse + Send + Sync + 'static,
    {
        Route {
            method,
            path: path.into(),
            handler: Arc::new(handler),
        }
    }
}

/// Options passed to [`mock_server`].
#[derive(Default)]
pub struct MockServerOpts {
    /// Optional fixed port — defaults to an OS-assigned ephemeral port so
    /// parallel tests do not clash.
    pub port: Option<u16>,
    /// Route table evaluated in registration order; first match wins.
    pub routes: Vec<Route>,
    /// Optional per-request timeout. `None` means hyper's default (no timeout).
    pub timeout: Option<Duration>,
}

impl MockServerOpts {
    /// Add a route — chainable builder helper.
    pub fn with_route(mut self, route: Route) -> Self {
        self.routes.push(route);
        self
    }

    /// Bind to a fixed port — chainable builder helper.
    pub fn with_port(mut self, port: u16) -> Self {
        self.port = Some(port);
        self
    }
}

/// Running mock server handle returned by [`mock_server`].
///
/// The server runs on a dedicated tokio multi-thread runtime (2 worker threads)
/// owned by the handle. The test thread stays sync — reqwest's blocking client
/// drives the wire — and dropping this handle shuts the runtime down so the
/// port is released deterministically.
pub struct MockServer {
    base_url: String,
    addr: SocketAddr,
    recorder: Arc<Mutex<Vec<RecordedRequest>>>,
    shutdown_tx: Option<oneshot::Sender<()>>,
    server_task: Option<JoinHandle<()>>,
    // Runtime must be dropped LAST — declared last so Drop order honours that.
    runtime: Option<Runtime>,
}

impl MockServer {
    /// Base URL clients should target (e.g. `http://127.0.0.1:54812`).
    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    /// Bound socket address (host + port).
    pub fn addr(&self) -> SocketAddr {
        self.addr
    }

    /// Bound port.
    pub fn port(&self) -> u16 {
        self.addr.port()
    }

    /// Snapshot of every request the server has received so far.
    pub fn recorded_requests(&self) -> Vec<RecordedRequest> {
        self.recorder
            .lock()
            .expect("kiwa mock_server recorder mutex poisoned")
            .clone()
    }

    /// Number of requests received so far.
    pub fn request_count(&self) -> usize {
        self.recorder
            .lock()
            .expect("kiwa mock_server recorder mutex poisoned")
            .len()
    }

    /// Stop the server explicitly. Idempotent — `Drop` calls this too.
    pub fn stop(&mut self) {
        if let Some(tx) = self.shutdown_tx.take() {
            // Receiver may already be gone if Drop fired between these calls;
            // ignore the error in that case.
            let _ = tx.send(());
        }
        if let (Some(task), Some(rt)) = (self.server_task.take(), self.runtime.as_ref()) {
            // Wait briefly for the task to drain; if it does not exit, abort.
            let _ = rt.block_on(async move {
                tokio::select! {
                    res = task => res.map(|_| ()).map_err(|e| e.to_string()),
                    _ = tokio::time::sleep(Duration::from_millis(500)) => Err("timeout".into()),
                }
            });
        }
    }
}

impl Drop for MockServer {
    fn drop(&mut self) {
        self.stop();
        // Drop the runtime explicitly here to release the port deterministically.
        // (Otherwise the field-order Drop would still run it, but doing it here
        // keeps the contract obvious.)
        if let Some(rt) = self.runtime.take() {
            rt.shutdown_background();
        }
    }
}

/// Spin up a hyper-based mock server on a background tokio runtime.
///
/// Blocks the calling thread until the server has bound a port, then returns a
/// [`MockServer`] handle. Tests drive the server with any sync HTTP client
/// (e.g. `reqwest::blocking::Client`) pointed at [`MockServer::base_url`] and
/// inspect [`MockServer::recorded_requests`] afterwards.
///
/// # Examples
///
/// ```no_run
/// use kiwa::integration::{mock_server, HttpMethod, MockResponse, MockServerOpts, Route};
///
/// let server = mock_server(MockServerOpts::default().with_route(Route::new(
///     HttpMethod::Get,
///     "/users",
///     |_req| MockResponse::json(b"[]".to_vec()),
/// )));
/// // … point a reqwest client at server.base_url(), then …
/// assert_eq!(server.request_count(), 0);
/// ```
pub fn mock_server(opts: MockServerOpts) -> MockServer {
    // Each server owns a dedicated multi-thread runtime so the test thread
    // can stay sync (reqwest::blocking spins its own runtime) without having
    // to drive ours. Two worker threads is enough for one mock endpoint and
    // keeps memory pressure low when many tests run in parallel.
    let runtime = tokio::runtime::Builder::new_multi_thread()
        .worker_threads(2)
        .enable_all()
        .build()
        .expect("kiwa mock_server failed to build tokio runtime");

    let recorder: Arc<Mutex<Vec<RecordedRequest>>> = Arc::new(Mutex::new(Vec::new()));
    let routes = Arc::new(opts.routes);
    let timeout = opts.timeout;

    // Bind synchronously so callers see a ready server (port known) on return.
    let bind_addr: SocketAddr = if let Some(port) = opts.port {
        ([127, 0, 0, 1], port).into()
    } else {
        ([127, 0, 0, 1], 0).into()
    };

    let (listener, addr) = runtime
        .block_on(async move {
            let listener = TcpListener::bind(bind_addr).await?;
            let addr = listener.local_addr()?;
            Ok::<_, std::io::Error>((listener, addr))
        })
        .expect("kiwa mock_server failed to bind TCP listener");

    let base_url = format!("http://{}", addr);

    let (shutdown_tx, mut shutdown_rx) = oneshot::channel::<()>();

    let recorder_for_task = Arc::clone(&recorder);
    let server_task = runtime.spawn(async move {
        loop {
            tokio::select! {
                _ = &mut shutdown_rx => break,
                accept = listener.accept() => {
                    let (stream, _peer) = match accept {
                        Ok(pair) => pair,
                        Err(_) => continue,
                    };
                    let io = TokioIo::new(stream);
                    let recorder_for_conn = Arc::clone(&recorder_for_task);
                    let routes_for_conn = Arc::clone(&routes);
                    let timeout_for_conn = timeout;
                    tokio::spawn(async move {
                        let service = service_fn(move |req: Request<Incoming>| {
                            let recorder = Arc::clone(&recorder_for_conn);
                            let routes = Arc::clone(&routes_for_conn);
                            async move {
                                let response = handle_request(req, recorder, routes, timeout_for_conn).await;
                                Ok::<_, Infallible>(response)
                            }
                        });
                        let _ = http1::Builder::new().serve_connection(io, service).await;
                    });
                }
            }
        }
    });

    MockServer {
        base_url,
        addr,
        recorder,
        shutdown_tx: Some(shutdown_tx),
        server_task: Some(server_task),
        runtime: Some(runtime),
    }
}

async fn handle_request(
    req: Request<Incoming>,
    recorder: Arc<Mutex<Vec<RecordedRequest>>>,
    routes: Arc<Vec<Route>>,
    request_timeout: Option<Duration>,
) -> Response<Full<Bytes>> {
    let method = req.method().clone();
    let path_and_query = req
        .uri()
        .path_and_query()
        .map(|pq| pq.as_str().to_string())
        .unwrap_or_else(|| req.uri().path().to_string());
    let path_only = req.uri().path().to_string();

    // Decode header values with to_str() at the boundary (invalid UTF-8
    // values are dropped) and delegate the two-view construction to the
    // shared recorder helper — the SSOT with kiwa::axum + kiwa::actix
    // (v1.6-5, Issue #611).
    let hint = req.headers().len();
    let pairs = req
        .headers()
        .iter()
        .filter_map(|(name, value)| value.to_str().ok().map(|v| (name.as_str(), v)));
    let (headers, headers_all) = crate::recorder::fold_headers(pairs, hint);

    // Collect the request body (with optional timeout) so the recorder sees the
    // full payload before we hand it to the route handler.
    let body_bytes = if let Some(t) = request_timeout {
        match tokio::time::timeout(t, req.into_body().collect()).await {
            Ok(Ok(c)) => c.to_bytes().to_vec(),
            Ok(Err(_)) | Err(_) => Vec::new(),
        }
    } else {
        match req.into_body().collect().await {
            Ok(c) => c.to_bytes().to_vec(),
            Err(_) => Vec::new(),
        }
    };

    let recorded = RecordedRequest {
        method: method.as_str().to_string(),
        path: path_and_query,
        headers,
        headers_all,
        body: body_bytes,
    };
    {
        let mut guard = recorder
            .lock()
            .expect("kiwa mock_server recorder mutex poisoned");
        guard.push(recorded.clone());
    }

    for route in routes.iter() {
        if route.method.matches(&method) && route.path == path_only {
            let mock_resp = (route.handler)(&recorded);
            return build_response(mock_resp);
        }
    }

    // No route matched — surface a 404 with a kiwa-flavoured body so tests
    // notice rogue requests.
    let body = format!(
        "kiwa mock_server: no route matched {} {}",
        method.as_str(),
        path_only,
    );
    Response::builder()
        .status(StatusCode::NOT_FOUND)
        .header("content-type", "text/plain; charset=utf-8")
        .body(Full::new(Bytes::from(body)))
        .expect("kiwa mock_server failed to build 404 response")
}

fn build_response(resp: MockResponse) -> Response<Full<Bytes>> {
    let status = StatusCode::from_u16(resp.status).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
    let mut builder = Response::builder().status(status);
    // headers_all is applied first; each value becomes a distinct header line
    // via http::Response::builder().header() (which append-appends when the
    // same key is added multiple times).
    for (k, values) in resp.headers_all.iter() {
        for v in values {
            builder = builder.header(k, v);
        }
    }
    // headers overrides any prior value on the same key so callers who only
    // fill `headers` keep the pre-v1.6 single-value behaviour on the wire.
    // We do this by removing prior entries in the builder headers_mut view
    // and re-adding a single value.
    for (k, v) in resp.headers.iter() {
        if let Some(hmap) = builder.headers_mut() {
            hmap.remove(k.as_str());
        }
        builder = builder.header(k, v);
    }
    builder
        .body(Full::new(Bytes::from(resp.body)))
        .unwrap_or_else(|_| {
            Response::builder()
                .status(StatusCode::INTERNAL_SERVER_ERROR)
                .body(Full::new(Bytes::from_static(
                    b"kiwa mock_server response build failed",
                )))
                .expect("kiwa mock_server fallback response build failed")
        })
}
